#  1.Write a function called rectangle that takes two integers m and n as arguments and prints out an m × n box consisting of asterisks.

In [1]:
def rectangle(m, n): # m is number of rows and n is number of coloumns 

    for i in range(m): # running loop to inialize rows
        for j in range(n): # running loop to inialize columns
            print("*"),
        print("")

rectangle(6,10)

* * * * * * * * * * 
* * * * * * * * * * 
* * * * * * * * * * 
* * * * * * * * * * 
* * * * * * * * * * 
* * * * * * * * * * 


# 2.Write a function called add_excitement that takes a list of strings and adds an exclamation point ( ! ) to the end of each string in the list. The program should modify the original list and not return anything.

In [3]:
def add_exitement():
    punctuations = "!"
    str = raw_input("Enter string: ") # takes input from the user

    # removes punctuation from the string
    no_punct = ""
    for char in str:
        if char not in punctuations:
            no_punct = no_punct + char

    print(no_punct) # displays the unpunctuated string

add_exitement() # calling function

Enter string: Hello! I am Ashikuzzaman!
Hello I am Ashikuzzaman


# 3.Write the same function except that it should not modify the original list and should instead return a new list.

In [4]:
def add_exitement():
    
    str = raw_input("Enter string: ") # takes a string as input
   
    print '\nNew String:'
    return str # returns the string

add_exitement() # calling function

Enter string: Hello! I am Ashikuzzaman!

New String:


'Hello! I am Ashikuzzaman!'

# 4.The digital root of a number n is obtained as follows: Add up the digits n to get a new number. Add up the digits of that to get another new number. Keep doing this until you get a number that has only one digit. That number is the digital root. For example, if n = 45893 , we add up the digits to get 4 + 5 + 8 + 9 + 3 = 29 . We then add up the digits of 29 to get 2 + 9 = 11 . We then add up the digits of 11 to get 1 + 1 = 2 . Since 2 has only one digit, 2 is our digital root. Write a function that returns the digital root of an integer n . [Note: there is a shortcut, where the digital root is equal to n mod 9, but do not use that here.]

In [5]:
def digi_root():
    
    n = input("Enter an integer: ")
    sum = 0
    while(n > 0 or sum > 9): # to find digital root, input must be greater than zero 
                             # and the sum of the digits must be greater than 9
            
        if(n == 0): # if input is 0, sum will be 0
            n = sum
            sum = 0
        # if input is grater than 0
        else:
            sum = sum + (n % 10)
            n = n / 10
    
    print "Digital Root is: ", sum

digi_root()

Enter an integer: 25371
Digital Root is:  9


Description:

Here n = 25371 is the given integer. First, we add digits of this integer(2+5+3+7+1=18). Then, each of the digits of 18 is added (1+8=9). The the outcome nine is a single digit. So, it is the digital root.

Calculation:

sum = (25371 % 10) = 1

n = (25371 / 10) = 2537

sum = 1 + (2537 % 10) = 8

n = (2537 / 10) = 253

sum = 8 + (253 % 10) = 11

n = (253 / 10) = 25

sum = 11 + (25 % 10) = 16

n = (25 / 10) = 2

sum = 16 + (2 % 10) = 18

n = (2 / 10) = 0

sum = 18 + (0 % 10) = 18

n = (0 / 10) = 0

Therefore,the Sum of Digits is 2+5+3+7+1 = 18 then 1+8 = 9. Here 9 is a single digit that is lower than 10, so the function stops calling and returns the output which is 9.

# 5.Write a function that takes an integer n and returns a random integer with exactly n digits. For instance, if n is 3, then 125 and 593 would be valid return values, but 093 would not because that is really 93, which is a two-digit number.

In [6]:
from random import randint # to generate random integer

def random_num():
    n = input("Enter the length of the integer: ") # input of the length of the integer
    
    start = 10 ** (n - 1) # starting point of n digit numbers
    end = (10 ** n) - 1 # end point of n digit numbers
    print '\nHere is a ', n, 'digit random number: '
    return randint(start, end) # returns a random number

random_num()

Enter the length of the integer: 3

Here is a  3 digit random number: 


692

Description:

Here, we use "randint" function from random module to generate a random number.

Then we take an input of the length of the integer.

Here, "start = 10 ** (n - 1)" means the starting point of n digit integers.

and, "end = (10 ** n) - 1" means the ending point of n digit integers.

If we input 3 that means the 3 digit integer length stars at 100 and ends at 999.

Then "return randint(start, end)" generates a random 3 digit integers and prints that integer.

# 6.Write a function called matches that takes two strings as arguments and returns how many matches there are between the strings. A match is where the two strings have the same character at the same index. For instance, ' python ' and ' path ' match in the first, third, and fourth characters, so the function should return 3.

In [7]:
def matches(str_1, str_2): # function name
  
    # set(str_1): set of characters of first string
    # set(str_2): set of characters of second string
    
    match = set(str_1) & set(str_2) # finds the matches between two strings  
     
    # prints the number of matched characters between the strings
    print 'Number of matched characters: ' +str(len(match))

matches('Robert','Hobart') # calling function with arguments

Number of matched characters: 4


Description:

In this function we are taking two strings as arguments. They are named "str_1" and "str_2".

Here, "str_1 = Rowan" and "str_2 = Ronald".

"set(str_1)" makes a set of all the characters in "str_1" which is {R, o, b, e, r, t}.

"set(str_2)" makes a set of all the characters in "str_2 which is {H, o, b, a, r, t}.

"match = set(str_1) & set(str_2)" finds the matches between the strings and stores them in "match" variable.

"+str(len(match))" calculates the total number of matches and prints them.

# 7.Write a function called primes that is given a number n and returns a list of the first n primes. Let the default value of n be 100.

In [8]:
def prime_num(n):
    while True:
        n += 1 # initializes every integer till n
        for i in range(2, n): # searches for prime number
            if n % i == 0: # checks if a number is prime or not
                break
        else:
            return n # returns prime number
def primes(N):
    n, count = 1,1
    while count <= N: # runs loop till n
        n = prime_num(n) # initializes the function "prime_num"
        yield n # returns a generator
        count += 1 # runs the loop every time by 1
N = input("How many prime numbers you want: ")
print 'List of', N, 'prime numbers: '
l = [x for x in primes(N)] # stores prime numbers till n
print l # returns all the prime numbers till n

How many prime numbers you want: 100
List of 100 prime numbers: 
[2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53, 59, 61, 67, 71, 73, 79, 83, 89, 97, 101, 103, 107, 109, 113, 127, 131, 137, 139, 149, 151, 157, 163, 167, 173, 179, 181, 191, 193, 197, 199, 211, 223, 227, 229, 233, 239, 241, 251, 257, 263, 269, 271, 277, 281, 283, 293, 307, 311, 313, 317, 331, 337, 347, 349, 353, 359, 367, 373, 379, 383, 389, 397, 401, 409, 419, 421, 431, 433, 439, 443, 449, 457, 461, 463, 467, 479, 487, 491, 499, 503, 509, 521, 523, 541]


# 8.Recall that if s is a string, then s.find( ' a ' ) will find the location of the first a in s . The problem is that it does not find the location of every a. Write a function called findall that given a string and a single character, returns a list containing all of the locations of that character in the string. It should return an empty list if there are no occurrences of the character in the string.

In [9]:
def findall(): # function name
    s = raw_input("Enter a string: ") # takes a string
    ch_loc = -1 
    print 'All the locations of a are: ',
    while True:
        ch_loc = s.find('a', ch_loc + 1) # s.find() is used to search "a" in a string 
                                         # find() is used to search in a string
            
        if ch_loc == -1: # when no "a" is found in the string
            break        # ends the program and returns empty list
        else:            # when "a" is found in the string
            yield ch_loc # returns a generator
            
for ch_loc in findall(): # runs for all the the locations of "a" in findall()
    print (ch_loc), # returns all the locations of "a"

Enter a string: My name is ashikuzzaman pias
All the locations of a are:  4 11 19 21 26


Description:

This function finds all the locations of "a" in a string.

Here "s.find' is used to find the locations of "a" in a string.

If no "a" is found in a string, it returns an empty list.

"yield" has been used as a generator that initializes whenever a location of "a" is found.